Tests/FluentAssertions.Specs/Collections/Data/DataColumnCollectionAssertionExtensionsSpecs.cs

```diff
@@ -0,0 +1,438 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Data;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+using FluentAssertions.Execution;
+
+using Xunit;
+using Xunit.Sdk;
+
+namespace FluentAssertions.Specs.Collections.Data
+{
+    public static class DataColumnCollectionAssertionExtensionsSpecs
+    {
+        private static DataColumn CreateTestDataColumn(int seed)
+        {
+            var column = new DataColumn("Column" + seed);
+
+            var random = new Random(seed);
+
+            column.DataType =
+                random.Next(4) switch
+                {
+                    0 => typeof(int),
+                    1 => typeof(string),
+                    2 => typeof(bool),
+                    _ => typeof(DateTime),
+                };
+
+            return column;
```

Comment:
What is the purpose of `CreateTestDataColumn`?

It looks like unnecessary complexity and indirection to the tests.

E.g. instead of having this in a test body
```cs
var dataTable = new DataTable();

for (int seed = 0; seed < 3; seed++)
{
    dataTable.Columns.Add(CreateTestDataColumn(seed));
}
```

this would suffice and avoid calling a helper method.
```cs
var dataTable = new DataTable();
dataTable.Columns.Add(new DataColumn());
dataTable.Columns.Add(new DataColumn());
dataTable.Columns.Add(new DataColumn());
```

The same comment applies to the other `private static Data* Create` helper methods.

Reply:
The `DataColumn`s have to have some properties initialized. At the very least, you have to tell them what data type they are and what their name is. It never occurred to me to try to add an uninitialized `DataColumn` to the collection.

I've just looked it up, and it looks like it'll alter the supplied `DataColumn` and give it a name if you haven't done that -- actually using pretty much the same scheme as `CreateTestDataColumn` uses, `Column1` then `Column2` then `Column3` etc. -- and if you don't assign to the data type, then every column will be of type `string`.

I'm of the opinion that it is more in line with tests not doing things implicitly and instead showing their functioning to explicitly assign a colmun name, and I am of the opinion that the tests are better if they exercise a variety of column data types instead of all columns being of type `string`.

Similar reasoning applies to the other helper methods:

* If `CreateTestDataTable` in `DataTableCollectionAssertionExtensionsSpecs` does not assign a name to the table, then the act of adding it to a `DataSet` will implicitly assign a name.
* Both that `CreateTestDataTable` and the `CreateTestDataTable` in `DataRowCollectionAssertionExtensionsSpecs` create a table that _actually has columns_ with a variety of data types. If you just do `new DataTable`, the resulting object is not typical of actual `DataTable`s, having no columns.
* The `AddTestDataRow` method in `DataRowCollectionAssertionExtensionsSpecs` populates rows with unique data. It is not a test of typical usage to test tables that have only a bunch of identical rows, especially if those rows have no columns.

In addition, the `BeSubsetOf` and `IntersectWith` functionality that was previously in this PR hasn't been discarded, it's just been moved to its own branch, and the tests for those _need_ rows that have unique data with respect to one another.

Reply:
Perhaps a small comment at the top of the file that explains that `System.Data` types have complex rules about the information they contain, and that the tests are intended to be working on somewhat "typical" specimens, would be helpful??

Reply:
For a real world `DataTable` I agree it does not make sense to initialize it without a data type.
In our opinion a unit test should only include the necessary parts.
Including code that is irrelevant to the test hides the important parts.

I agree that _some_ tests should exercise a variety of column data types, but unless you're exercising an FA assertion that acts on column data types, these "some" tests belong to your test code or the MS tests implementing System.Data types.

Take e.g. `When_two_collections_have_the_same_number_of_columns_it_should_succeed`.

```cs
// Arrange
var firstDataTable = new DataTable();
var secondDataTable = new DataTable();

for (int seed = 0; seed < 3; seed++)
{
    firstDataTable.Columns.Add(CreateTestDataColumn(seed));
    secondDataTable.Columns.Add(CreateTestDataColumn(seed + 10));
}

// Act & Assert
firstDataTable.Columns.Should().HaveSameCount(secondDataTable.Columns);
```

I just verified that you don't need to helper method or loop to create a passing test.

```cs
// Arrange
var firstDataTable = new DataTable();
firstDataTable.Columns.Add(new DataColumn());
firstDataTable.Columns.Add(new DataColumn());
firstDataTable.Columns.Add(new DataColumn());

var secondDataTable = new DataTable();
secondDataTable.Columns.Add(new DataColumn());
secondDataTable.Columns.Add(new DataColumn());
secondDataTable.Columns.Add(new DataColumn());

// Act & Assert
firstDataTable.Columns.Should().HaveSameCount(secondDataTable.Columns);
```

With regard to `BeSubsetOf` etc. please an issue first, about those before spending time on a PR.
I really don't want to waste your time if we decide that they don't fit into FA.

Reply:
The code is already written for them. There's not a question of doing additional work -- the work is already done.

Reply:
To be clear, I have been applying feedback from the existing PRs to them as I go. For instance, the set operations are written using the fluent syntax in the latest commit, and all failure assertions test the exception message. It's been incremental updates at the same time as the work on this PR, and it's already in that state. So, I think the most expedient way of creating issues for them is to create PRs for them. The code is already there. :-)

Reply:
Mmkay, well, it felt important in these tests to cover a variety of data types in columns. The tests don't depend on it being specific types, and aren't testing specific types, but having everything default to `string` feels like the test isn't actually exercising typical use cases.

Reply:
This is complete.

Reply:
Thanks for the follow-up commits, it did simplify things 👍 

There are still three helper methods that I'd like to get rid of.
* `CreateTestDataTableWithColumns`
* `CreateTestDataTable`
* `AddTestDataRow`

Reply:
I personally think this is a case where the tests are easier to reason about with the helper methods, but I've pushed a commit that eliminates them.